text/0000-canonical-metadata.md

```diff
@@ -0,0 +1,1232 @@
+* Replicated type system in java and javascript that is maintained manually
+* Dependency(reference) analysis is done outside of metadata gathering
+* Change management to accomodate new metadata is cumbersome because the implementation is spread across two repos
+
+# Detailed Design
+
+## Format
+Metadata shape of the various objects we wish to analyse will be represented using in [JSON Schema](https://json-schema.org/) format. This is also augmented with a TypeScript definition of the shape.
+
+[JSON](https://tools.ietf.org/html/rfc7159) is the IETF approved standard for representing objects. [JSON Schema](https://json-schema.org/) is on the path to become an IETF approved standard for describing JSON.
+
+This format was also influenced by standardization choices made with in Salesforce UI platform.
+
+## Entities that will be analyzed
+These are the entities that will be analyzed and metadata gathered about:
+* HTML Template file
+    * Custom element references
+        * Tag name
+        * Attributes
+        * Slot content - default and named slots
+    * Static resource
+        * Image urls
+        * Javascript resource
+        * CSS resource
+    * Dynamic components
+    * Nodes with special directives
+        * for:each
+        * if
+        * iterator
+        * key
+        * lwc:dom="manual"
+    * Event handlers
+    * Slots
+* Javascript file
+    * Component classes
+        * Inheritance
+        * Public & private properties
+        * Decorators
+        * getters/setters
+        * Public methods
+    * Imports
+        * Builtin modules
+        * External modules
+        * Salesforce scoped modules
+        * Other modules(components and libraries)
+        * Relative imports
+        * Dynamic imports
+    * Exports
+        * Default exports
+        * Named exports
+        * Aggregating modules
+    * JSDoc
+        * Class
+        * Properties
+        * Public methods
+    * Programmatically generated events
+        * type
+        * options: composed, bubbles
+    * Programmatically fetched static resources
+        * `<script>`
+        * `<img>`
+        * `<link>`
+* CSS file
+    * Tokens
+    * Static resources
+
+## Metadata shape
+
+### TypeScript
+
+#### Bundle metadata shape
+```ts
+interface BundleMetadata {
+    version: string;
+    name?: string;
+    namespace?: string;
+    // Define the type of module that is represented by the given metadata object.
+    moduleType?: "HTML" | "CSS" | "Javascript";
+
+    /**
+     * For a component bundle, this property is a reference to the main component class in the bundle.
+     * This class must be the default export of the javascript file. The file must be named the same
+     * as the bundle
+     */
+    defaultComponentClass?: Class;
```

Comment:
since this property is optional, i suggest we include 'exports', which would be an array of all javascript exports coming from the entry file, whether it is a class, function or a const